arm: debug: Add GDB stub for aarch32

marek-vedral opened the pull request:
Add implementation of GDB stub for 32-bit ARM. It has been
tested only on the Zynq-7000 SoC and I would like to get any feedback
from others.

The stub still has these issues:

- To implement single stepping, it uses instruction address mismatch
  breakpoint, as recommended in ARMv7 reference. The breakpoint control
  register is configured (the state control fields) for the "PL0, Supervisor
  and System modes only" option. Otherwise the breakpoint would also halt the
  processor in abort mode, in which the stub loop runs. Zephyr kernel
  runs in the system mode. This works well until the kernel enables
  interrupts, as interrupt handlers typically run in Supervisor mode.
  Single stepping therefore sometimes "catches" a handler instead of the
  next application instruction. I have not tried User mode, because
  Cortex-A SoCs do not appear to have the ARCH_HAS_USERSPACE flag.

    @wentasah
    Cc: Michal Sojka <michal.sojka@cvut.cz>
    Signed-off-by: Marek Vedral <marek.vedral@gmail.com>

marek-vedral committed:
arm: debug: Add GDB stub for aarch32

This commit adds implementation of GDB stub for 32-bit ARM. It has been
tested only on the Zynq-7000 SoC and I would like to get any feedback
from others.

The stub still has these issues:

- To implement single stepping, it uses instruction address mismatch
  breakpoint, as recommended in ARMv7 reference. The breakpoint control
  register is configured (the state control fields) for the "PL0,
  Supervisor and System modes only" option. Otherwise the breakpoint
  would also halt the processor in abort mode, in which the stub loop
  runs. Zephyr kernel runs in the system mode. This works well until the
  kernel enables interrupts, as interrupt handlers typically run in
  Supervisor mode. Single stepping therefore sometimes "catches" a
  handler instead of the next application instruction. I have not tried
  User mode, because Cortex-A SoCs do not appear to have the
  ARCH_HAS_USERSPACE flag.

Cc: Michal Sojka <michal.sojka@cvut.cz>
Signed-off-by: Marek Vedral <marek.vedral@gmail.com>

marek-vedral committed:
drivers/timer/arm_arch_timer: Fix cycles overflow with GDB stub

If GDBSTUB is enabled and the kernel runs in tickless mode, the timer
must not convert the delta cycles to a 32-bit data type (cycle_diff_t in
this case). The delta_ticks variable would overflow and the next timeout
would be set before the current timestamp, thus generating an interrupt
right after leaving the handler. As a result, the system would receive
tens of thousands of interrupts per second and would not boot.

Cc: Michal Sojka <michal.sojka@cvut.cz>
Signed-off-by: Marek Vedral <marek.vedral@gmail.com>

dcpleung commented on subsys/debug/gdbstub.c:
```diff
@@ -36,7 +36,7 @@ LOG_MODULE_REGISTER(gdbstub);
 #define GDB_ERROR_OVERFLOW  "E22"
 
-static bool not_first_start;
+static int not_first_start;
```
The problem described doesn't seem right. Maybe it's an toolchain issue?

marek-vedral replied:
The file is now restored back to the original. I tested it again with the bool data type and it seems to work fine.

microbuilder commented:
I'm not sure many of the reviewers here will have a Zync-7000 to test with (I don't). Could this be made to work with `qemu_cortex_a9`, which is the same core used on the Zync-7000?

marek-vedral commented:
Yes, it does seem to work on the emulator, except for single stepping. The emulator seems to ignore the mismatch breakpoint and continues the execution. Do you know whether QEMU also emulates these debug registers?

I tried to run the `samples/subsys/debug/gdbstub` program. The property `zephyr,gdbstub-uart = &uart1;` needs to be added to the devicetree of `qemu_cortex_a9` and the same option `-serial tcp:127.0.0.1:5678,server` as for `qemu_x86` must be specified in CMakeList.txt.

microbuilder commented:
> Yes, it does seem to work on the emulator, except for single stepping. The emulator seems to ignore the mismatch breakpoint and continues the execution. Do you know whether QEMU also emulates these debug registers?

I'm not sure about the internals myself, but if we have an easy to reproduce qemu sample I can ping the people who would know around me. See point below to that effect.

> I tried to run the `samples/subsys/debug/gdbstub` program. The property `zephyr,gdbstub-uart = &uart1;` needs to be added to the devicetree of `qemu_cortex_a9` and the same option `-serial tcp:127.0.0.1:5678,server` as for `qemu_x86` must be specified in CMakeList.txt.

Since we'll probably want some sort of basic testing for CI for this, could you add those changes for the qemu target in a separate commit, and we can figure out how to put together a basic test scenario from there?

marek-vedral committed:
samples: debug: Add support for qemu_cortex_a9 in GDB stub sample

This commit adds a devicetree overlay file and an extra condition in the
CMakeLists.txt file to allow remote debugging for the qemu_cortex_a9
board.

Signed-off-by: Marek Vedral <marek.vedral@gmail.com>

marek-vedral commented:
> Since we'll probably want some sort of basic testing for CI for this, could you add those changes for the qemu target in a separate commit, and we can figure out how to put together a basic test scenario from there?

Added a new commit with those changes.

microbuilder commented:
> Added a new commit with those changes.

Thanks. The sample folder is missing some files to be a valid sample, but if you can add either some notes in the sample's readme, plus a useful main.c source file and project config file, or update this PR with some instructions on testing this, I can kick the tires in on it.

MaureenHelm commented:
@povergoing can you take a look?

dcpleung commented on arch/Kconfig:
```diff
@@ -29,6 +29,8 @@ config ARM
 	bool
 	select ARCH_IS_SET
 	select ARCH_SUPPORTS_COREDUMP if CPU_CORTEX_M
+	# GDBSTUB has not yet been tested on Cortex M or R SoCs
+	select ARCH_HAS_GDBSTUB if CPU_AARCH32_CORTEX_A
```
Maybe put this inside `arch/arm/core/aarch32/Kconfig` under `CPU_AARCH32_CORTEX_A`? M and R can do the same to their kconfigs when support is added.

dcpleung commented on arch/Kconfig:
```diff
@@ -496,6 +498,8 @@ config IRQ_OFFLOAD_NESTED
 
 config EXTRA_EXCEPTION_INFO
 	bool "Collect extra exception info"
+	# GDB on ARM needs the etxra registers
+	default y if GDBSTUB && ARM
```
I assume those extra (type: etxra) registers are mandatory, this should be done via `select` (under `arch/arm/core/aarch32/Kconfig` as mentioned above). Using `default` here means that it can be overridden via command line or defconfig downstream of arch.

povergoing commented on arch/arm/core/aarch32/cortex_a_r/fault.c:
```diff
@@ -191,6 +196,12 @@ bool z_arm_fault_undef_instruction(z_arch_esf_t *esf)
+#if defined(CONFIG_GDBSTUB)
+	z_gdb_entry(esf, GDB_EXCEPTION_INVALID_INSTRUCTION);
+	/* Might not be fatal if GDB stub placed it in the code. */
+	return false;
```
As I saw you `return false` after `z_gdb_entry`, would you like to elaborate on the details if there is a real fault instead of a gdb exception? You are going to send all faults to the gdb clients?

povergoing reviewed (changes requested):
Some minor typos and issues need to be fixed.

also, link to #60031 in case there is an impact. @SgrrZhf @ithinuel  would you also take a look?

marek-vedral commented:
Rebased the commits to make the requested changes.

> Maybe put this inside `arch/arm/core/aarch32/Kconfig` under `CPU_AARCH32_CORTEX_A`? M and R can do the same to their kconfigs when support is added.

> I assume those extra (type: etxra) registers are mandatory, this should be done via `select` (under `arch/arm/core/aarch32/Kconfig` as mentioned above). Using `default` here means that it can be overridden via command line or defconfig downstream of arch.

Both options are now selected under CPU_AARCH32_CORTEX_A.

> Can you please use macros to define the registers, e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/arch/arm/aarch32/cortex_a_r/lib_helpers.h

Those functions appear to only modify the CP15 registers.

> do you mean `memset(buf, "x", 8);`?

Yes, I fixed it in the commit.

> What's the meaning of `ret = 8`, here? Why not `int ret = 8` at the beginning? Can you also explain a bit the meaning of the 8? It's the default size of the function read?

I don't remember why I chose 8. The return value could be anything except 0. It is now set at the beginning of the function to 4 to simply indicate that four bytes were read.

> I found that all `arch_gdb_reg_read/writexxx`s share almost the same logic, would you like to refine them to make them more compact if it's not too difficult?

I am not sure how to make them more elegant. The logic is needed due to the way the registers are stored in GDB packets.

> Can this one also be set by arch/soc/board specifically instead of `default xxx if ARM`

Yes, I moved it to the aarch32 Kconfig file.

> As I saw you `return false` after `z_gdb_entry`, would you like to elaborate on the details if there is a real fault instead of a gdb exception? You are going to send all faults to the gdb clients?

Yes, all faults are redirected to GDB. Even if there is a real fault, the user probably should be notified. I don't know whether it would be worth to add some logic to analyze the instructions (to check which ones are GDB exceptions and which ones are not).

povergoing commented on drivers/timer/arm_arch_timer.c:
```diff
@@ -58,7 +66,7 @@ static void arm_arch_timer_compare_isr(const void *arg)
 	uint64_t curr_cycle = arm_arch_timer_count();
 	uint64_t delta_cycles = curr_cycle - last_cycle;
-	uint32_t delta_ticks = (cycle_diff_t)delta_cycles / CYC_PER_TICK;
+	uint32_t delta_ticks = TO_CYCLE_DIFF(delta_cycles) / CYC_PER_TICK;
```
What about removing the `TO_CYCLE_DIFF`, just using `uint32_t delta_ticks = (cycle_diff_t) (delta_cycles / CYC_PER_TICK);`

povergoing replied:
Then, any differences between `uint32_t delta_ticks = delta_cycles / CYC_PER_TICK` and `uint32_t delta_ticks = (uint32_t) delta_cycles / CYC_PER_TICK` as `delta_ticks` is a `uint32_t`

marek-vedral replied:
Oh, sorry, you are right. I didn't notice that the parenthesis encloses the whole division in your previous comment.

I would still keep the macro, because we want to cast the `delta_cycles` variable to `uint32_t` if possible to improve performance.

povergoing replied:
Okay, fair enough

povergoing commented:
> Those functions appear to only modify the CP15 registers.

Right, I didn't notice that, would you mind commenting with the register name, as it's hard for someone to find the register in the ARM spec just with the register code?

>I am not sure how to make them more elegant. The logic is needed due to the way the registers are stored in GDB packets.

Okay, not a big problem.

> Yes, all faults are redirected to GDB. Even if there is a real fault, the user probably should be notified. I don't know whether it would be worth to add some logic to analyze the instructions (to check which ones are GDB exceptions and which ones are not).

Noted, BTW, you have tested it on the qemu_cortex_a7, right?

marek-vedral commented:
> > Those functions appear to only modify the CP15 registers.
> 
> Right, I didn't notice that, would you mind commenting with the register name, as it's hard for someone to find the register in the ARM spec just with the register code?

Allright, I thought that was clear, but I can make a new commit.

> 
> > I am not sure how to make them more elegant. The logic is needed due to the way the registers are stored in GDB packets.
> 
> Okay, not a big problem.
> 
> > Yes, all faults are redirected to GDB. Even if there is a real fault, the user probably should be notified. I don't know whether it would be worth to add some logic to analyze the instructions (to check which ones are GDB exceptions and which ones are not).
> 
> Noted, BTW, you have tested it on the qemu_cortex_a7, right?

No, I tested it on qemu_cortex_a9. The Cortex A7 emulator is not yet in Zephyr.

SgrrZhf commented:
> also, link to #60031 in case there is an impact. @SgrrZhf @ithinuel would you also take a look?

Yes, it seems that I have new files that need to take care of.  Don't worry, it doesn't matter.

microbuilder commented:
@povergoing @SgrrZhf Do you have any further feedback on this preventing you from approving the PR?

povergoing reviewed (commented):
Oh sorry for the late response, I am okay with this

povergoing commented:
> @povergoing @SgrrZhf Do you have any further feedback on this preventing you from approving the PR?

Except there is an unrelated CI failure

microbuilder commented:
@marek-vedral Approved this since the CI failures are unrelated, but can you please rebase to see if that solves the CI failures? Thanks.

povergoing commented:
@marek-vedral since there is a CI failure plus a conflict, would you mind rebasing the PR?

marek-vedral commented:
> @marek-vedral Approved this since the CI failures are unrelated, but can you please rebase to see if that solves the CI failures? Thanks.

Rebased on the main ranch.

SgrrZhf commented on arch/arm/core/aarch32/gdbstub.c:
```diff
@@ -0,0 +1,225 @@
+	__asm__ volatile("mcr p14, 0, %0, c0, c0, 5" ::"r"(reg_val) :);
+}
+
+size_t arch_gdb_reg_readall(struct gdb_ctx *ctx, uint8_t *buf, size_t buflen)
```
Can you fix this warning message
```
warning: declaration of 'ctx' shadows a global declaration [-Wshadow]
```

SgrrZhf commented:
I did a simple test locally and the functionality is well and just some GCC warning messages that should be address.

microbuilder commented:
> I did a simple test locally and the functionality is well and just some GCC warning messages that should be addressed.

@marek-vedral Do you mind doing a final update to the PR to take into account the change requests here? Thanks.

marek-vedral commented:
> > I did a simple test locally and the functionality is well and just some GCC warning messages that should be addressed.
> 
> @marek-vedral Do you mind doing a final update to the PR to take into account the change requests here? Thanks.

Yes, I renamed the function parameters.

microbuilder commented:
Still a lot of failures here. Do you have time to look through and sort some of them out to be able to get this merged in eventually? Thanks.

microbuilder commented:
> > Still a lot of failures here. Do you have time to look through and sort some of them out to be able to get this merged in eventually? Thanks.
> 
> I tried rebasing on the main branch again. Most of the checks are successful, not sure why the two remaining checks fail each time.

This needs to merge first: https://github.com/zephyrproject-rtos/zephyr/pull/62431

microbuilder commented:
> This needs to merge first: #62431

Should pass now if you rebase.

marek-vedral commented:
> > This needs to merge first: #62431
> 
> Should pass now if you rebase.

Thanks for the notification, I rebased on the main branch again.

SgrrZhf commented:
It has conflicts now. Would you please rebase it again? @marek-vedral

marek-vedral commented:
> It has conflicts now. Would you please rebase it again? @marek-vedral

Rebased on the main branch, there should not be conflicts anymore.

SgrrZhf commented:
I have tested it again and I always get stuck in the GDB command line after I type `n` and the application is running to the end and won't stop. This sample (`samples/subsys/debug/gdbstub`) works well on the `qemu_x86` board well. Could you check and test it again?

marek-vedral commented:
> I have tested it again and I always get stuck in the GDB command line after I type `n` and the application is running to the end and won't stop. This sample (`samples/subsys/debug/gdbstub`) works well on the `qemu_x86` board well. Could you check and test it again?

We discussed the issue with @microbuilder, see the earlier comments. I am not sure whether QEMU also emulates the debug registers, which would allow single stepping.

carlescufi commented:
@marek-vedral please rebase

marek-vedral commented:
> @marek-vedral please rebase

Rebased on main

microbuilder commented:
@povergoing Do you mind giving this another look? Thanks.

microbuilder commented:
@carlocaione Do you have time to give this a review?
